File: power_state.py
from __future__ import annotations

import os
import threading
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any

class PowerStateBase(ABC):
    """Abstract base class defining the public interface for PowerState.

    This class defines all externally-used methods. Subclasses provide the
    implementation, while internal helper methods (prefixed with _) remain
    private to the implementation.
    """

    @abstractmethod
    def wait_for_change(self, from_state: str, timeout: int) -> str:
        """Wait for the state to change and return the new state."""
        pass

    @abstractmethod
    def request_state_change(self, new_state: str) -> None:
        """Request a change to a new state: "on", "off", or "disconnected"."""
        pass

class PowerStateEmulator(PowerStateBase):
    """Emulates power state changes for testing purposes."""
    def __init__(self) -> None:
        """Initialize instance state and start the connection watchdog thread."""
        self._state: str = "disconnected"
        self._condition = threading.Condition()
        self._watchdog_thread = threading.Thread(target=self.connection_watchdog, daemon=True)
        self._watchdog_thread.start()

    def wait_for_change(self, from_state: str, timeout: int) -> str:
        with self._condition:
            if self._state == from_state:
                self._condition.wait(timeout=timeout)
            return self._state

    def request_state_change(self, new_state: str) -> None:
        with self._condition:
            if self._state == "loading":
                    return  # Ignore state change requests while loading
            if new_state != self._state:
                self._state = "loading"
                self._condition.notify_all()  # Notify that we're now loading
                # Simulate loading time
                self._condition.wait(timeout = 1)
                if self._state == "loading":
                    self._state = new_state  # Transition to the new state after loading
                    self._condition.notify_all()

    def connection_watchdog(self) -> None:
        """Simulate a connection watchdog that reconnects the disconnected state."""
        while True:
            with self._condition:
                if self._state != "disconnected":
                    self._condition.wait(timeout=10)  # Wait for state change or timeout
                    continue
                print("connection_watchdog: Detected disconnected state, waiting for reconnection...")
                self._condition.wait(timeout=10)
                if self._state != "disconnected":
                    print("connection_watchdog: Reconnection detected, state is now", self._state)
                    continue  # State changed, check again immediately
                print("connection_watchdog: No reconnection detected after 10 seconds, simulating reconnection...")
                self._state = "loading"
                self._condition.notify_all()
                self._condition.wait(timeout=1)  # Simulate loading time
                print("connection_watchdog: Transitioning to 'on' state")
                self._state = "on"
                self._condition.notify_all()

import paho.mqtt.client as mqtt
import json
BROKER = "localhost"  # Change to your MQTT broker address
PORT = 1883
KEEPALIVE = 60
CLIENT_ID = "on-off controller"

class PowerStateMQTT(PowerStateBase):
    """power state driven by MQTT messages."""
    def __init__(self) -> None:
        self._host = os.getenv("MQTT_HOST", "localhost")
        self._port = int(os.getenv("MQTT_PORT", "1883"))
        self._tasmota_id = os.getenv("TASMOTA_ID", "XXXXXX")
        self._state = "disconnected"
        self._disconnect_count = 0
        self._loading_count = 0
        self._condition: threading.Condition = threading.Condition()
        self._mqtt_client: Any = None  # Placeholder for MQTT client
        self._power_state: str = "unknown"
        self._power_state_update_time: float = 0.0

        self._lwt_topic = f"tele/tasmota_{self._tasmota_id}/LWT"
        self._power_topic = f"stat/tasmota_{self._tasmota_id}/POWER"
        self._power_command_topic = f"cmnd/tasmota_{self._tasmota_id}/POWER"

        """Start the MQTT manager thread after initialization."""
        self.mqtt_thread = threading.Thread(target=self.mqtt_manager, daemon=True)
        self.mqtt_thread.start()

    def wait_for_change(self, from_state: str, timeout: int) -> str:
        with self._condition:
            if self._state == from_state:
                self._condition.wait(timeout=timeout)
            return self._state

    def request_state_change(self, new_state: str) -> None:
        with self._condition:
            if self._state not in {"on", "off"}:
                print(f"request_state_change: Ignoring state change request to '{new_state}' because current state is '{self._state}'")
                return  # Ignore state change requests while loading or connecting

            self._state = "loading"
            self._loading_count += 1
            self._condition.notify_all()  # Notify that we're now loading

            if new_state in {"on", "off"}:
                if self._mqtt_client is not None:
                    try:
                        self._mqtt_client.publish(self._power_command_topic, new_state.upper())
                        print(f"request_state_change: Published MQTT message to change state to '{new_state}'")
                        return # Assume state will change based on MQTT message, wait for confirmation in mqtt_manager
                    except Exception as e:
                        print("Failed to publish MQTT message:", e)

            self._cleanup_client()  # If we can't publish, treat it as a disconnection
            with self._condition:
                self._state = "disconnected"
                self._disconnect_count += 1
                print(f"request_state_change: Failed to publish MQTT message, setting state to 'disconnected'")
                self._condition.notify_all()

    def mqtt_manager(self) -> None:
        time.sleep(2)  # Brief delay to allow main thread to start
        """MQTT manager to handle connection and messages."""
        with self._condition:
            self._state = "loading"
            self._loading_count += 1
            print ("mqtt_manager: state set to 'loading', count: ", self._loading_count)
            self._condition.notify_all()  # Notify that we're now loading

        need_reconnect = True
        need_disconnect = False
        while True:
            if need_reconnect:
                print("mqtt_manager: Attempting to start MQTT client...")
                self._start_client()
                need_reconnect = False

            if need_disconnect:
                print(f"mqtt_manager: Attempting to clean up MQTT client at address {hex(id(self._mqtt_client))}...")
                self._cleanup_client()
                with self._condition:
                    self._state = "disconnected"
                    self._disconnect_count += 1
                    self._condition.notify_all()
                need_disconnect = False

            with self._condition:
                previous_state = self._state
                previous_disconnect_count = self._disconnect_count
                previous_loading_count = self._loading_count
                self._condition.wait(timeout=10)  # Wait for state change or timeout
                if self._state == "disconnected" and previous_state == "disconnected" and self._disconnect_count == previous_disconnect_count:
                    print(f"mqtt_manager: Detected prolonged disconnected state (count {self._disconnect_count}), will attempt to reconnect...")
                    self._state = "loading"
                    self._loading_count += 1
                    self._condition.notify_all()
                    need_reconnect = True
                    continue
                if self._state == "loading" and previous_state == "loading" and self._loading_count == previous_loading_count:
                    print(f"mqtt_manager: Detected prolonged loading state (count {self._loading_count}), disconnecting")
                    need_disconnect = True
                    continue

    def _start_client(self) -> None:
        """Start the MQTT client and connect to the broker."""
        self._mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, client_id=CLIENT_ID, reconnect_on_failure=False)
        print (f"_start_client: Created MQTT client with id {hex(id(self._mqtt_client))}")
        self._mqtt_client.on_connect = self._on_connect
        self._mqtt_client.on_disconnect = self._on_disconnect
        self._mqtt_client.on_message = self._on_message
        try:
            print(f"Attempting to connect to MQTT broker at {self._host}:{self._port}...")
            connect_start_time = time.time()
            self._mqtt_client.connect(self._host, self._port, KEEPALIVE)
            print("MQTT client connected successfully after {:.2f} seconds".format(time.time() - connect_start_time))
            self._mqtt_client.loop_start()
        except Exception as e:
            elapsed_time = time.time() - connect_start_time
            print(f"_start_client: MQTT connection failed after {elapsed_time:.2f} seconds:", e)
            with self._condition:
                self._state = "disconnected"
                self._disconnect_count += 1
                print(f"state set to 'disconnected' for client {hex(id(self._mqtt_client))} in _start_client due to connection failure")
                self._condition.notify_all()

    def _cleanup_client(self) -> None:
        client = self._mqtt_client
        self._mqtt_client = None
        if client is None:
            return
        try:
            client.disconnect()
        except Exception:
            pass
        try:
            client.loop_stop()
        except Exception:
            pass

    def _on_connect(self, client, userdata, flags, rc, properties=None):
        # check if connection was successful
        if rc != 0:
            print("MQTT connection failed with result code "+str(rc))
            # Destroy the client to trigger a clean reconnect.
            self._cleanup_client()
            with self._condition:
                self._state = "disconnected"
                self._disconnect_count += 1
                print(f"state set to 'disconnected' for client {hex(id(self._mqtt_client))} in _on_connect due to failed connection")
                self._condition.notify_all()
            return
        # Subscribe to command topic
        client.subscribe(self._lwt_topic)
        client.subscribe(self._power_topic)
        print(f"Subscribed to MQTT topics: {self._lwt_topic}, {self._power_topic}")
        client.publish(self._power_command_topic)  # Request current state immediately

    def _on_disconnect(self, client, userdata, flags, rc, properties=None):
        if not client is self._mqtt_client:
            print(f"Received on_disconnect for an old client {hex(id(client))}, ignoring")
            return
        with self._condition:
            self._state = "disconnected"
            self._disconnect_count += 1
            print(f"state set to 'disconnected' for client {hex(id(self._mqtt_client))} in _on_disconnect")
            self._condition.notify_all()

    def _on_message(self, client, userdata, msg):
        topic = msg.topic
        payload = msg.payload.decode().strip()
        print(f"Received MQTT message: {topic} {payload}")
        cleanup_needed = False
        with self._condition:
            if topic.endswith("/LWT"):
                if payload == "Online":
                    print("MQTT LWT indicates device is online")
                else:
                    print("MQTT LWT indicates device is offline")
                    cleanup_needed = True
            elif topic.endswith("/POWER"):
                if payload in {"ON", "OFF"}:
                    print(f"MQTT POWER state changed to {payload}")
                    self._state = payload.lower()
                    self._condition.notify_all()
                return

        if cleanup_needed:
            with self._condition:
                if self._state != "loading":  # Only set to loading if we're not already in the middle of a state change
                    self._state = "loading"
                    self._loading_count += 1
                    print("mqtt_manager: state set to 'loading', count: ", self._loading_count)

                    self._condition.notify_all()
            self._cleanup_client()  # on_disconnect will be called which will set state to disconnected and trigger reconnect logic in mqtt_manager

File: power_state.go
package main

import (
	"context"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

const (
	brokerKeepAlive = 60 * time.Second
	clientIDPrefix  = "on-off-controller"
)

type PowerState interface {
	WaitForChange(ctx context.Context, fromState string, timeout time.Duration) string
	RequestStateChange(newState string)
	Close()
}

type PowerStateEmulator struct {
	mu       sync.Mutex
	notifyCh chan struct{}
	state    string
	stopCh   chan struct{}
}

func NewPowerStateEmulator() *PowerStateEmulator {
	p := &PowerStateEmulator{
		notifyCh: make(chan struct{}),
		state:    "disconnected",
		stopCh:   make(chan struct{}),
	}
	go p.connectionWatchdog()
	return p
}

func (p *PowerStateEmulator) WaitForChange(ctx context.Context, fromState string, timeout time.Duration) string {
	deadline := time.Now().Add(timeout)
	for {
		p.mu.Lock()
		state := p.state
		if state != fromState {
			p.mu.Unlock()
			return state
		}
		ch := p.notifyCh
		remaining := time.Until(deadline)
		p.mu.Unlock()

		if remaining <= 0 {
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		}

		select {
		case <-ch:
		case <-time.After(remaining):
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		case <-ctx.Done():
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		}
	}
}

func (p *PowerStateEmulator) RequestStateChange(newState string) {
	p.mu.Lock()
	if p.state == "loading" {
		p.mu.Unlock()
		return
	}
	if newState == p.state {
		p.mu.Unlock()
		return
	}
	p.state = "loading"
	p.signalLocked()
	p.mu.Unlock()

	timer := time.NewTimer(1 * time.Second)
	defer timer.Stop()

	select {
	case <-timer.C:
		p.mu.Lock()
		if p.state == "loading" {
			p.state = newState
			p.signalLocked()
		}
		p.mu.Unlock()
	case <-p.stopCh:
	}
}

func (p *PowerStateEmulator) Close() {
	select {
	case <-p.stopCh:
	default:
		close(p.stopCh)
	}
}

func (p *PowerStateEmulator) connectionWatchdog() {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			p.mu.Lock()
			if p.state == "disconnected" {
				logln("connection_watchdog: No reconnection detected after 10 seconds, simulating reconnection...")
				p.state = "loading"
				p.signalLocked()
				p.mu.Unlock()

				time.Sleep(1 * time.Second)

				p.mu.Lock()
				p.state = "on"
				p.signalLocked()
				p.mu.Unlock()
				continue
			}
			p.mu.Unlock()
		case <-p.stopCh:
			return
		}
	}
}

func (p *PowerStateEmulator) signalLocked() {
	close(p.notifyCh)
	p.notifyCh = make(chan struct{})
}

type PowerStateMQTT struct {
	clientID string

	host    string
	port    int
	tasmota string

	lwtTopic          string
	powerTopic        string
	powerCommandTopic string

	mu              sync.Mutex
	notifyCh        chan struct{}
	state           string
	disconnectCount int
	loadingCount    int
	client          mqtt.Client

	stopCh chan struct{}
	doneCh chan struct{}
}

func NewPowerStateMQTT() *PowerStateMQTT {
	tasmotaID := getenv("TASMOTA_ID", "XXXXXX")
	hostname, _ := os.Hostname()
	defaultClientID := fmt.Sprintf("%s-%s-%d", clientIDPrefix, strings.ReplaceAll(hostname, " ", "-"), os.Getpid())
	p := &PowerStateMQTT{
		clientID:          getenv("MQTT_CLIENT_ID", defaultClientID),
		host:              getenv("MQTT_HOST", "localhost"),
		port:              atoiDefault(getenv("MQTT_PORT", "1883"), 1883),
		tasmota:           tasmotaID,
		lwtTopic:          fmt.Sprintf("tele/tasmota_%s/LWT", tasmotaID),
		powerTopic:        fmt.Sprintf("stat/tasmota_%s/POWER", tasmotaID),
		powerCommandTopic: fmt.Sprintf("cmnd/tasmota_%s/POWER", tasmotaID),
		notifyCh:          make(chan struct{}),
		state:             "disconnected",
		stopCh:            make(chan struct{}),
		doneCh:            make(chan struct{}),
	}
	go p.mqttManager()
	return p
}

func (p *PowerStateMQTT) WaitForChange(ctx context.Context, fromState string, timeout time.Duration) string {
	deadline := time.Now().Add(timeout)
	for {
		p.mu.Lock()
		state := p.state
		if state != fromState {
			p.mu.Unlock()
			return state
		}
		ch := p.notifyCh
		remaining := time.Until(deadline)
		p.mu.Unlock()

		if remaining <= 0 {
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		}

		select {
		case <-ch:
		case <-time.After(remaining):
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		case <-ctx.Done():
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		case <-p.stopCh:
			p.mu.Lock()
			state = p.state
			p.mu.Unlock()
			return state
		}
	}
}

func (p *PowerStateMQTT) RequestStateChange(newState string) {
	p.mu.Lock()
	if p.state != "on" && p.state != "off" {
		logf("request_state_change: Ignoring state change request to '%s' because current state is '%s'\n", newState, p.state)
		p.mu.Unlock()
		return
	}

	p.state = "loading"
	p.loadingCount++
	p.signalLocked()
	client := p.client
	topic := p.powerCommandTopic
	p.mu.Unlock()

	if (newState == "on" || newState == "off") && client != nil {
		token := client.Publish(topic, 0, false, strings.ToUpper(newState))
		if token.WaitTimeout(5*time.Second) && token.Error() == nil {
			logf("request_state_change: Published MQTT message to change state to '%s'\n", newState)
			return
		}
		if err := token.Error(); err != nil {
			logln("Failed to publish MQTT message:", err)
		}
	}

	p.cleanupClient()
	p.mu.Lock()
	p.state = "disconnected"
	p.disconnectCount++
	logln("request_state_change: Failed to publish MQTT message, setting state to 'disconnected'")
	p.signalLocked()
	p.mu.Unlock()
}

func (p *PowerStateMQTT) Close() {
	select {
	case <-p.stopCh:
	default:
		close(p.stopCh)
	}
	<-p.doneCh
}

func (p *PowerStateMQTT) mqttManager() {
	defer close(p.doneCh)

	p.mu.Lock()
	p.state = "loading"
	p.loadingCount++
	logln("mqtt_manager: state set to 'loading', count:", p.loadingCount)
	p.signalLocked()
	p.mu.Unlock()

	needReconnect := true
	needDisconnect := false

	for {
		select {
		case <-p.stopCh:
			p.cleanupClient()
			return
		default:
		}

		if needReconnect {
			logln("mqtt_manager: Attempting to start MQTT client...")
			p.startClient()
			needReconnect = false
		}

		if needDisconnect {
			logln("mqtt_manager: Attempting to clean up MQTT client...")
			p.cleanupClient()
			p.mu.Lock()
			p.state = "disconnected"
			p.disconnectCount++
			p.signalLocked()
			p.mu.Unlock()
			needDisconnect = false
		}

		p.mu.Lock()
		prevState := p.state
		prevDisconnectCount := p.disconnectCount
		prevLoadingCount := p.loadingCount
		waitCh := p.notifyCh
		p.mu.Unlock()

		select {
		case <-waitCh:
		case <-time.After(10 * time.Second):
		case <-p.stopCh:
			p.cleanupClient()
			return
		}

		p.mu.Lock()
		if p.state == "disconnected" && prevState == "disconnected" && p.disconnectCount == prevDisconnectCount {
			logf("mqtt_manager: Detected prolonged disconnected state (count %d), will attempt to reconnect...\n", p.disconnectCount)
			p.state = "loading"
			p.loadingCount++
			p.signalLocked()
			needReconnect = true
			p.mu.Unlock()
			continue
		}
		if p.state == "loading" && prevState == "loading" && p.loadingCount == prevLoadingCount {
			logf("mqtt_manager: Detected prolonged loading state (count %d), disconnecting\n", p.loadingCount)
			needDisconnect = true
			p.mu.Unlock()
			continue
		}
		p.mu.Unlock()
	}
}

func (p *PowerStateMQTT) startClient() {
	opts := mqtt.NewClientOptions()
	opts.AddBroker(fmt.Sprintf("tcp://%s:%d", p.host, p.port))
	opts.SetClientID(p.clientID)
	opts.SetKeepAlive(brokerKeepAlive)
	opts.SetAutoReconnect(false)
	opts.SetOnConnectHandler(func(c mqtt.Client) {
		if token := c.Subscribe(p.lwtTopic, 0, nil); token.Wait() && token.Error() != nil {
			logln("Subscribe error for LWT topic:", token.Error())
		}
		if token := c.Subscribe(p.powerTopic, 0, nil); token.Wait() && token.Error() != nil {
			logln("Subscribe error for POWER topic:", token.Error())
		}
		logf("Subscribed to MQTT topics: %s, %s\n", p.lwtTopic, p.powerTopic)
		c.Publish(p.powerCommandTopic, 0, false, "")
	})
	opts.SetConnectionLostHandler(func(c mqtt.Client, err error) {
		p.mu.Lock()
		if p.client != c {
			p.mu.Unlock()
			logln("Received on_disconnect for an old client, ignoring. Reason:", err)
			return
		}
		p.state = "disconnected"
		p.disconnectCount++
		logln("state set to 'disconnected' in on_disconnect. Reason:", err)
		p.signalLocked()
		p.mu.Unlock()
	})
	opts.SetDefaultPublishHandler(func(c mqtt.Client, msg mqtt.Message) {
		p.onMessage(c, msg)
	})

	client := mqtt.NewClient(opts)
	p.mu.Lock()
	p.client = client
	p.mu.Unlock()

	logf("Attempting to connect to MQTT broker at %s:%d...\n", p.host, p.port)
	connectStart := time.Now()
	token := client.Connect()
	if !token.WaitTimeout(10*time.Second) || token.Error() != nil {
		elapsed := time.Since(connectStart).Seconds()
		err := token.Error()
		if err == nil {
			err = fmt.Errorf("timeout")
		}
		logf("_start_client: MQTT connection failed after %.2f seconds: %v\n", elapsed, err)
		p.mu.Lock()
		p.state = "disconnected"
		p.disconnectCount++
		p.signalLocked()
		p.mu.Unlock()
		p.cleanupClient()
		return
	}
	logf("MQTT client connected successfully after %.2f seconds\n", time.Since(connectStart).Seconds())
}

func (p *PowerStateMQTT) cleanupClient() {
	p.mu.Lock()
	client := p.client
	p.client = nil
	p.mu.Unlock()
	if client != nil && client.IsConnected() {
		client.Disconnect(250)
	}
}

func (p *PowerStateMQTT) onMessage(client mqtt.Client, msg mqtt.Message) {
	topic := msg.Topic()
	payload := strings.TrimSpace(string(msg.Payload()))
	logf("Received MQTT message: %s %s\n", topic, payload)

	cleanupNeeded := false
	p.mu.Lock()
	if strings.HasSuffix(topic, "/LWT") {
		if payload == "Online" {
			logln("MQTT LWT indicates device is online")
		} else {
			logln("MQTT LWT indicates device is offline")
			cleanupNeeded = true
		}
	} else if strings.HasSuffix(topic, "/POWER") {
		if payload == "ON" || payload == "OFF" {
			logf("MQTT POWER state changed to %s\n", payload)
			p.state = strings.ToLower(payload)
			p.signalLocked()
		}
		p.mu.Unlock()
		return
	}
	p.mu.Unlock()

	if cleanupNeeded {
		p.mu.Lock()
		if p.state != "loading" {
			p.state = "loading"
			p.loadingCount++
			logln("mqtt_manager: state set to 'loading', count:", p.loadingCount)
			p.signalLocked()
		}
		p.mu.Unlock()
		p.cleanupClient()
	}
}

func (p *PowerStateMQTT) signalLocked() {
	close(p.notifyCh)
	p.notifyCh = make(chan struct{})
}

func getenv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func atoiDefault(s string, fallback int) int {
	var v int
	if _, err := fmt.Sscanf(s, "%d", &v); err != nil {
		return fallback
	}
	return v
}

func logf(format string, args ...any) {
	fmt.Printf(time.Now().Format("2006-01-02 15:04:05.000 ")+format, args...)
}

func logln(args ...any) {
	all := append([]any{time.Now().Format("2006-01-02 15:04:05.000")}, args...)
	fmt.Println(all...)
}
